src/test/java/com/auth0/utils/tokens/SignatureVerifierTest.java

```diff
@@ -0,0 +1,213 @@
+        exception.expectCause(isA(AlgorithmMismatchException.class));
+
+        SignatureVerifier verifier = SignatureVerifier.forHS256("secret");
+        verifier.verifySignature(RS_JWT);
+    }
+
+    @Test
+    public void failsWhenAlgorithmHS256IsNotExpected() throws Exception {
+        exception.expect(IdTokenValidationException.class);
+        exception.expectMessage("Token signed with an unexpected algorithm");
+        exception.expectCause(isA(AlgorithmMismatchException.class));
+
+        SignatureVerifier verifier = SignatureVerifier.forRS256(getRSProvider(RS_PUBLIC_KEY));
+
+        verifier.verifySignature(HS_JWT);
+    }
+
+    @Test
+    public void succeedsWithValidSignatureHS256Token() {
+        SignatureVerifier verifier = SignatureVerifier.forHS256("secret");
+        DecodedJWT decodedJWT = verifier.verifySignature(HS_JWT);
+
+        assertThat(decodedJWT, notNullValue());
+    }
+
+    @Test
+    public void succeedsAndIgnoresExpiredTokenException() {
+        SignatureVerifier verifier = SignatureVerifier.forHS256("secret");
+        DecodedJWT decodedJWT = verifier.verifySignature(EXPIRED_HS_JWT);
+
+        assertThat(decodedJWT, notNullValue());
+    }
+
+    @Test
+    public void failsWithInvalidSignatureHS256Token() {
+        exception.expect(IdTokenValidationException.class);
+        exception.expectMessage("Invalid token signature");
+
+        SignatureVerifier verifier = SignatureVerifier.forHS256("badsecret");
+        verifier.verifySignature(HS_JWT);
+    }
+
+    @Test
+    public void succeedsWithValidSignatureRS256Token() throws Exception {
+        SignatureVerifier verifier = SignatureVerifier.forRS256(getRSProvider(RS_PUBLIC_KEY));
+        DecodedJWT decodedJWT = verifier.verifySignature(RS_JWT);
+
+        assertThat(decodedJWT, notNullValue());
+    }
+
+    @Test
+    public void failsWithInvalidSignatureRS256Token() throws Exception {
+        exception.expect(IdTokenValidationException.class);
+        exception.expectMessage("Invalid token signature");
+        SignatureVerifier verifier = SignatureVerifier.forRS256(getRSProvider(RS_PUBLIC_KEY_BAD));
+        DecodedJWT decodedJWT = verifier.verifySignature(RS_JWT);
+
+        assertThat(decodedJWT, notNullValue());
+    }
+
+    @Test
+    public void failsWhenErrorGettingPublicKey() {
+        exception.expect(IdTokenValidationException.class);
+        exception.expectCause(isA(PublicKeyProviderException.class));
+        exception.expectMessage("Error retrieving public key");
+
+        SignatureVerifier verifier = SignatureVerifier.forRS256(new PublicKeyProvider() {
+            @Override
+            public RSAPublicKey getPublicKeyById(String keyId) throws PublicKeyProviderException {
+                throw new PublicKeyProviderException("error");
+            }
+        });
+        verifier.verifySignature(RS_JWT);
+    }
+
+    @Test
+    public void failsWhenErrorGettingPublicKeyAndHasNestedExceptionCause() {
```

Comment:
this case doesn't really need to be tested in this class. When you use a verifier, you are just checking if an exception is thrown. You don't care how that exception was created, if it had or not a nested exception, etc 

Our custom exceptions, because they just pass things up to the constructor and don't have any custom logic I wouldn't care much in code-covering them 👍

Reply:
Yeah, was kinda chasing coverage metrics to get it green. I don't love it either so will remove.